firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+            for func_index in indices:
+                arg = loopy.GlobalArg("coeff"+str(get), shape=coeff_shape_list[func_index], dtype="double")
+                args.append(arg)
+                get += 1
+        except ValueError:
+            pass
+
+    # Add all other cofficients
+    for v, k in builder.extra_coefficients:
+        args.append(loopy.GlobalArg(k,
+                                    shape=builder.args_extents[k],
+                                    dtype="double"))
+
+    # Arg for is exterior (==0)/interior (==1) facet or not
+    if builder.needs_cell_facets:
+        args.append(loopy.GlobalArg(builder.cell_facets_arg,
+                                    shape=(builder.num_facets, 2),
+                                    dtype=np.int8))
+
+    if builder.needs_mesh_layers:
+        args.append(loopy.TemporaryVariable("layer", shape=(), dtype=np.int32, address_space=loopy.AddressSpace.GLOBAL))
+
+    ############
+    # TODO: preprocessing of gem for removing unneccesary component tensors
+    # print("not peprocessed",traversed_gem_expr_dag)
+    # traversed_gem_expr_dag = impero_utils.preprocess_traversedgem(traversed_gem_expr_dag[0])
+    # print("preprocessed",traversed_gem_expr_dag)
+    # print(traversed_gem_expr_dag[1])
+    # traversed_gem_expr_dag=traversed_gem_expr_dag[1]
+
+    # glue assignments to return variable
+    assignments = list(zip(ret_vars, traversed_gem_expr_dag))
+    impero_c = impero_utils.compile_gem(assignments, (), remove_zeros=False)
+
+    # Part B: impero_c to loopy
+    precision = 12
+    loopy_outer = generate_loopy(impero_c, args, precision, "double", "loopy_outer")
+    return loopy_outer
+
+
+# TODO: those should got into firedrake loopy!
+# STAGE 2c: register external function calls
+# the get_*_callable replaces the according callable
+# with the c-function which is defined in the preamble
+def get_inv_callable(loopy_merged):
+    class INVCallable(loopy.ScalarCallable):
+        def __init__(self, name, arg_id_to_dtype=None, 
+                     arg_id_to_descr=None, name_in_target=None):
+
+            super(INVCallable, self).__init__(name,
+                  arg_id_to_dtype=arg_id_to_dtype,
+                  arg_id_to_descr=arg_id_to_descr)
+
+            self.name = name
+            self.name_in_target = name_in_target
+
+        def with_types(self, arg_id_to_dtype, kernel, callables_table):
+            for i in range(0, len(arg_id_to_dtype)):
+                if i not in arg_id_to_dtype or arg_id_to_dtype[i] is None:
+                    # the types provided aren't mature enough to specialize the
+                    # callable
+                    return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                            callables_table)
+
+            mat_dtype = arg_id_to_dtype[0].numpy_dtype
+            name_in_target = "inverse_"
+
+            from loopy.types import NumpyType
```

Comment:
Local import (see comment below)

Reply:
Adressed in 7aa9b698a1722b7f18f5b3f23332dbb95f064226